climada/engine/impact.py

```diff
@@ -48,6 +47,410 @@
+
+        The dimension of the imp_mat variable must be compatible with the
+        exposures and hazard objects.
+
+        Parameters
+        ----------
+        exposures : climada.entity.Exposures
+            exposure used to compute imp_mat
+        impf_set: climada.entity.ImpactFuncSet
+            impact functions set used to compute imp_mat
+        hazard : climada.Hazard
+            hazard used to compute imp_mat
+        imp_mat : sparse.csr_matrix, optional
+            matrix num_events x num_exp with impacts.
+            Default is an empty matrix.
+
+        Returns
+        -------
+        None.
+
+        """
+
+        imp_mat = sparse.csr_matrix(np.empty((0, 0))) if imp_mat is None else imp_mat
+        self.exposures = exposures
+        self.impfset = impfset
+        self.hazard = hazard
+        self.imp_mat = imp_mat
+        self.n_exp_pnt = self.exposures.gdf.shape[0]
+        self.n_events = self.hazard.size
+
+    @property
+    def deductible(self):
+        """
+        Deductible from the exposures
+
+        Returns
+        -------
+        np.array
+            The deductible per exposure point
+
+        """
+        return self.exposures.gdf['deductible']
+
+    @property
+    def cover(self):
+        """
+        Cover from the exposures
+
+        Returns
+        -------
+        np.array
+            The cover per exposure point
+
+        """
+        return self.exposures.gdf['cover']
+
+    def impact(self, save_mat=True):
+        """Compute the impact of a hazard on exposures.
+
+        Parameters
+        ----------
+        save_mat : bool
+            if true, save the total impact matrix (events x exposures)
+
+        Examples
+        --------
+            >>> haz = Hazard.from_mat(HAZ_DEMO_MAT)  # Set hazard
+            >>> impfset = ImpactFuncSet.from_excel(ENT_TEMPLATE_XLS) # Set impact functions
+            >>> exp = Exposures(pd.read_excel(ENT_TEMPLATE_XLS)) # Set exposures
+            >>> impcalc = ImpactCal(exp, impfset, haz)
+            >>> imp = impcalc.insured_impact()
+            >>> imp.aai_agg
+
+        Note
+        ----
+        Deductible and/or cover values in the exposures are ignored.
+        """
+        impf_col = self.exposures.get_impf_column(self.hazard.haz_type)
+        exp_gdf = self.minimal_exp_gdf(impf_col)
+        LOGGER.info('Calculating impact for %s assets (>0) and %s events.',
+                    self.n_events, self.n_events)
+        imp_mat_gen = self.imp_mat_gen(exp_gdf, impf_col)
+        if save_mat:
+            self.imp_mat = self.stitch_impact_matrix(imp_mat_gen)
+            at_event, eai_exp, aai_agg = self.risk_metrics()
+        else:
+            at_event, eai_exp, aai_agg = self.stitch_risk_metrics(imp_mat_gen)
+        return Impact.from_eih(
+            self.exposures, self.impfset, self.hazard,
+            at_event, eai_exp, aai_agg, self.imp_mat
+            )
+
+    def insured_impact(self, save_mat=False):
```

Comment:
This method shares a lot of instructions with `impact()`. I think it would be best to separate these parts into one or multiple private methods to avoid repeated code.

Reply:
How would you do that? There are already separated into the simplest methods I could think of. Except maybe the blob from `if save_mat` on. How would you call such a method?

Reply:
Great, exactly what I was thinking of 👌